base/utf.jl

```diff
@@ -0,0 +1,917 @@
+    @inbounds while pos < len
+        ch, pos = next(str, pos)
+        totalchar += 1
+        if ch > 0x7f
+            if ch < 0x100
+                num2byte += 1
+                flags |= UTF_LATIN1
+            elseif ch < 0x800
+                num2byte += 1
+                flags |= UTF_UNICODE2
+            elseif ch > 0xffff
+                (ch > 0x10ffff) && utf_errfunc(UTF_ERR_INVALID, pos, ch)
+                num4byte += 1
+            elseif !is_surrogate_char(ch)
+                num3byte += 1
+            elseif is_surrogate_lead(ch)
+                pos == len && utf_errfunc(UTF_ERR_MISSING_SURROGATE, pos, ch)
+                # next character *must* be a trailing surrogate character
+                ch, pos = next(str, pos)
+                !is_surrogate_trail(ch) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, ch)
+                num4byte += 1
+                (options & UTF_NO_SURROGATES) != 0 && utf_errfunc(UTF_ERR_SURROGATE, pos, ch)
+                flags |= UTF_SURROGATE
+            else
+                utf_errfunc(UTF_ERR_NOT_LEAD, pos, ch)
+            end
+        end
+    end
+    totalchar, flags | (num3byte == 0 ? 0 : UTF_UNICODE3) | (num4byte == 0 ? 0 : UTF_UNICODE4), num4byte, num3byte, num2byte
+end
+
+# Get the rest of a UTF-8 character from an array, update the position
+macro get_utf8_2!(str, pos, ch) ; quote ; (($(esc(ch)) & 0x1f) << 6) | ($(esc(str))[$(esc(pos)) += 1] & 0x3f) ; end ; end
+macro get_utf8_3!(str, pos, ch) ; quote ; ($(esc(pos)) += 2 ; (($(esc(ch)) & 0xf) << 12) | (UInt32($(esc(str))[$(esc(pos))-1] & 0x3f) << 6) | ($(esc(str))[$(esc(pos))] & 0x3f)) ; end ; end
+macro get_utf8_4!(str, pos, ch) ; quote ; ($(esc(pos)) += 3 ; (($(esc(ch)) & 0x7) << 18) | (UInt32($(esc(str))[$(esc(pos))-2] & 0x3f) << 12) | (UInt32($(esc(str))[$(esc(pos))-1] & 0x3f) << 6) | ($(esc(str))[$(esc(pos))] & 0x3f)) ; end ; end
+
+# Get the trailing surrogate character in UTF-8 from an array, update the position
+macro get_utf8_surr!(str, pos)  ; quote ; ($(esc(pos)) += 3 ; ((UInt32($(esc(str))[$(esc(pos))-2] & 0xf) << 12) | (UInt32($(esc(str))[$(esc(pos))-1] & 0x3f) << 6) | ($(esc(str))[$(esc(pos))] & 0x3f))) ; end ; end
+
+# Output a character as UTF-8, update the position
+macro output_utf8_2!(buf, out, ch) ; quote ; $(esc(buf))[$(esc(out)) += 1] = 0xc0 | ($(esc(ch)) >>> 6) ;  $(esc(buf))[$(esc(out)) += 1] = 0x80 | ($(esc(ch)) & 0x3f) ; end ; end
+macro output_utf8_3!(buf, out, ch) ; quote ; $(esc(buf))[$(esc(out)) += 1] = 0xe0 | (($(esc(ch)) >>> 12) & 0x3f) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | (($(esc(ch)) >>> 6) & 0x3f) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | ($(esc(ch)) & 0x3f) ; end ; end
+macro output_utf8_4!(buf, out, ch) ; quote ; $(esc(buf))[$(esc(out)) += 1] = 0xf0 | ($(esc(ch)) >>> 18) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | (($(esc(ch)) >>> 12) & 0x3f) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | (($(esc(ch)) >>> 6) & 0x3f) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | ($(esc(ch)) & 0x3f) ; end ; end
```

Comment:
These were originally very short one-line macros in C, but in Julia they got rather large... I'm still having problems figuring out how to write macros well in Julia, it seems I need `$(esc(varname))` everywhere, which makes things long and hard to read...  There's got to be a better way... any advice?